### PR TITLE
#1611 Regression in anti-flipping mechanism for mouselook camera

### PR DESCRIPTION
**Pull request by LLGuru**

### CODE REVIEW

**Comment by akleshchev** [comment]
I rezzed a Kart 1.0, rode it and when I loke up camera flips to the other orientation. I think that isn't intended?

But Kart and tank are working fine when on a vertical surface. Was able to raise tank's barrel without issues.

**Review by akleshchev — APPROVED** [review]

File: indra/newview/llagent.cpp [hunk 1]
```diff
@@ -1459,69 +1459,64 @@ LLVector3 LLAgent::getReferenceUpVector()
...
-            check_viewer_camera = true;
+            // Camera sight relative to agent frame (focus - offset)
+            LLVector3 camera_offset(gAgentCamera.getCameraOffsetInitial());
+            LLVector3 camera_focus(gAgentCamera.getFocusOffsetInitial());
+            LLVector3 camera_sight(camera_focus - camera_offset);
+            // 2D projection of the camera sight to the XZ plane
+            LLVector2 camera_sight_2d_vert(1, camera_sight[VZ]);
+            camera_sight_2d_vert.normalize();
+            // Cosine of the 2D angle between initial camera At and X axis (in the XZ plane)
+            F32 camera_sight_to_at_2d_vert_cos = camera_sight_2d_vert * LLVector2(LLVector3::x_axis);
```
Comment (akleshchev) [inline indra/newview/llagent.cpp:1500]: I'm not sure XZ plane is the right chose. 'plane' is relative. Ex: If you are flying a vehicle vertically or upside down...

**Review by marchcat — APPROVED** [review]

File: indra/newview/llagent.cpp (hunk 1 above)
Comment (LLGuru) [reply indra/newview/llagent.cpp:1500]: I've meant an XZ plane in agent coordinate system, not in world
In the THIRD_PERSON mode the viewer camera and the agent camera have the constant vertical angle between each other
For example, if it is 10 degrees, then when agent looks up 80 degrees over horizont, the viewer camera looks at zenith
This code just calculates this vertical angle between the viewer camera and the agent camera

**Review by marchcat — COMMENTED** [review]
This looks much cleaner now.

**Comment by akleshchev** [comment]
Personaly I keep a 'do not touch' policy with math classes. They work and they work well and had worked for years, changing anything, especially en masse always has a chance to break something.

**Comment by marchcat** [comment]
Right, it's usually no point in fixing what isn't broken. Not to mention that it can (and often does) introduce some new bugs.

**Pull request merged**